src/wallet/api.rs

```diff
@@ -1255,63 +1260,97 @@ impl Wallet {
                 .as_ref()
                 .map(|addr| addr.clone().assume_checked().to_string())
                 .unwrap_or_else(|| format!("script_{}", utxo.script_pub_key));
+            address_groups
+                .entry(address_str)
+                .or_default()
+                .push((utxo, spend_info));
+        }
+
+        let (mut grouped_addresses, single_addresses): (Vec<_>, Vec<_>) = address_groups
+            .into_values()
+            .partition(|group| group.len() > 1);
+
+        // Sort by value
+        grouped_addresses
+            .sort_by_key(|group| group.iter().map(|(u, _)| u.amount.to_sat()).sum::<u64>());
+
+        // Mojo's Ultra-Compressed: Single loop with let binding
+        let (selected_utxos, selected_total, selected_weight) = {
+            let mut result_utxos = Vec::new();
+            let mut result_total = 0u64;
+            let mut result_weight = 0u64;
+
+            for group in grouped_addresses {
+                let group_total: u64 = group.iter().map(|(u, _)| u.amount.to_sat()).sum();
+                let group_weight: u64 = group
+                    .iter()
+                    .map(|(_, spend_info)| {
+                        32 + 4 + 4 + 1 + spend_info.estimate_witness_size() as u64
+                    })
+                    .sum();
 
-            match address_groups
-                .iter_mut()
-                .find(|(addr, _)| addr == &address_str)
-            {
-                Some(group) => group.1.push((utxo.clone(), spend_info.clone())),
-                None => {
-                    address_groups.push((address_str, vec![(utxo.clone(), spend_info.clone())]))
+                // Inline fee calculation (3 lines)
+                let tx_weight = 43 + result_weight + group_weight + target_weight.to_wu();
+                let estimated_fee = calculate_fee(tx_weight / 4, feerate as f32)
+                    .map_err(|e| WalletError::General(format!("Fee calculation failed: {e}")))?;
+
+                result_total += group_total;
+                result_weight += group_weight;
+                result_utxos.extend(group);
+
+                // Check if we have enough (either single group or accumulated)
+                if result_total >= target_amount + estimated_fee {
+                    log::info!(
+                        "Address grouping: Selected {} UTXOs (total: {} sats, target+fee: {} sats)",
+                        result_utxos.len(),
+                        result_total,
+                        target_amount + estimated_fee
+                    );
+                    return Ok(result_utxos);
                 }
             }
-        }
-
-        let grouped_utxos = address_groups
-            .into_iter()
-            .map(|(_, utxos)| utxos)
-            .collect::<Vec<_>>();
+            (result_utxos, result_total, result_weight)
+        };
 
-        let output_groups = grouped_utxos
+        // Run coin selection on single addresses
```

Comment:
mention that grouped selection didn't work, thus running regular coinselection on the remaining amount.